In [1]:
%load_ext autoreload
%autoreload 2

In [3]:
import datetime
import pandas as pd
import numpy as np
from functools import partial

from utils.datafetch import *
from utils.vectorized_funs import *
from utils.datapipe import *

> We'll need some way of loading our possible tickets.

> We'll have some information for AMEX, NASDAQ and NYSE on one csv each in the config folder

In [4]:
rr = load_exchangesinfos()

amex has 400 tickers.
nasdaq has 3081 tickers.
nyse has 3299 tickers.


Final dataset has 6780 records
(NOK) Final dataset has duplicate tickers (6779 distinct)



> We'll look into the duplicate tickers when applicable

In [5]:
_r = duplicate_tickers(rr, ign_exchange=True)
_r

Symbol                               Name     MarketCap  \
2716    MSG  The Madison Square Garden Company  5.837670e+09   
2922    MSG  The Madison Square Garden Company  5.837670e+09   

                 Sector             Industry exchange ignoring_exchange  
2716  Consumer Services  Television Services   nasdaq             False  
2922  Consumer Services  Television Services     nyse              True

> in this case the ticker is the same company appearing in two exchanges, so we'll drop one of them

In [6]:
rr = rr.drop([2922])

In [7]:
## this time this should be empty...

duplicate_tickers(rr, ign_exchange=True)

Empty DataFrame
Columns: [Symbol, Name, MarketCap, Sector, Industry, exchange, ignoring_exchange]
Index: []

> lets have a look at your datareader wrapper

In [8]:
fetch_quotes("AMZN", from_date=datetime.datetime(2016,12,1))

Open        High         Low       Close    Volume  \
Date                                                                   
2016-12-01  752.409973  753.369995  738.030029  743.650024   4666000   
2016-12-02  743.400024  748.489990  736.700012  740.340027   3561300   
2016-12-05  745.000000  761.489990  742.000000  759.359985   4314700   
2016-12-06  763.989990  768.239990  757.250000  764.719971   3794700   
2016-12-07  764.549988  770.419983  755.820007  770.419983   3684900   
2016-12-08  771.869995  773.789978  765.190002  767.330017   3189600   
2016-12-09  770.000000  770.250000  765.340027  768.659973   2470900   
2016-12-12  766.400024  766.890015  757.200012  760.119995   2963900   
2016-12-13  764.960022  782.460022  762.000000  774.340027   5285300   
2016-12-14  778.250000  780.859985  762.809998  768.820007   5454800   
2016-12-15  766.280029  769.099976  760.309998  761.000000   3801900   
2016-12-16  765.000000  765.130005  754.000000  757.770020   4848200   
2016-12-19  758.890015  770.500000  756.159973  766.000000   3113200   
2016-12-20  768.650024  774.390015  767.710022  771.219971   2703600   
2016-12-21  770.000000  771.219971  765.700012  770.599976   2044600   
2016-12-22  768.119995  771.210022  763.020020  766.340027   2543600   
2016-12-23  764.549988  766.500000  757.989990  760.590027   1981600   
2016-12-27  763.400024  774.650024  761.200012  771.400024   2638700   
2016-12-28  776.250000  780.000000  770.500000  772.130005   3301000   
2016-12-29  772.400024  773.400024  760.849976  765.150024   3158300   
2016-12-30  766.469971  767.400024  748.280029  749.869995   4139400   
2017-01-03  757.919983  758.760010  747.700012  753.669983   3521100   
2017-01-04  758.390015  759.679993  754.200012  757.179993   2510500   
2017-01-05  761.549988  782.400024  760.260010  780.450012   5830100   
2017-01-06  782.359985  799.440002  778.479980  795.989990   5986200   
2017-01-09  798.000000  801.770020  791.770020  796.919983   3440100   
2017-01-10  796.599976  798.000000  789.539978  795.900024   2558400   
2017-01-11  793.659973  799.500000  789.510010  799.020020   2992800   
2017-01-12  800.309998  814.130005  799.500000  813.640015   4873900   
2017-01-13  814.320007  821.650024  811.400024  817.140015   3791900   
...                ...         ...         ...         ...       ...   
2017-02-01  829.210022  833.780029  824.940002  832.349976   3797000   
2017-02-02  836.590027  842.489990  828.260010  839.950012   6541700   
2017-02-03  806.719971  818.299988  804.000000  810.200012  10838900   
2017-02-06  809.799988  810.719971  803.000000  807.640015   3808200   
2017-02-07  809.309998  816.159973  807.500000  812.500000   3461500   
2017-02-08  812.690002  821.479980  812.500000  819.710022   2848300   
2017-02-09  821.599976  825.000000  819.710022  821.359985   2468400   
2017-02-10  823.820007  828.000000  822.849976  827.460022   2425800   
2017-02-13  831.619995  843.000000  828.549988  836.530029   4156500   
2017-02-14  837.000000  838.309998  831.450012  836.390015   2787900   
2017-02-15  834.000000  842.809998  832.820007  842.700012   2950600   
2017-02-16  841.840027  845.000000  839.380005  844.140015   2708100   
2017-02-17  842.000000  847.270020  840.729980  845.070007   3029800   
2017-02-21  848.840027  857.979980  847.250000  856.440002   3490100   
2017-02-22  856.950012  858.429993  852.179993  855.609985   2613400   
2017-02-23  857.570007  860.859985  848.000000  852.190002   3454600   
2017-02-24  844.690002  845.809998  837.750000  845.239990   3676400   
2017-02-27  842.380005  852.500000  839.669983  848.640015   2702300   
2017-02-28  851.450012  854.090027  842.049988  845.039978   2777700   
2017-03-01  853.049988  854.830017  849.010010  853.080017   2752000   
2017-03-02  853.080017  854.820007  847.280029  848.909973   2129200   
2017-03-03  847.200012  851.989990  846.270020  849.880005   1941100   
2017-03-06  845.229980  848.489990  841.119995  846.609985

> Lets see how we could prepare an initial data load

In [11]:
tkrs = dict()
errs = dict()
itr = None

for tkr in rr.iloc[0:5]["Symbol"].values.tolist():
    itr = fetch_quotes(tkr)
    if itr is None:
        errs[tkr] = itr
        print("ERROR for {}".format(tkr))
    else:
        tkrs[tkr] = itr
        print("Received {}".format(tkr))


Received XXII
Received FAX
Received IAF
Received CH
ERROR for ABE           


In [9]:
_r = initial_dataload(["AAPL", "TWTR", "XXII"])

(1/3) Recv. and Stored AAPL
(2/3) Recv. and Stored TWTR
(3/3) Recv. and Stored XXII
Took 0:00:00.882025


> lets try with the ticker dataframe now

In [11]:
OKs = ["OK" for x in _r["OK"]]
NOKs = ["NOK" for x in _r["NOK"]]

o = pd.DataFrame([OKs, _r["OK"]]).T
n = pd.DataFrame([NOKs, _r["NOK"]]).T

d = pd.concat([o, n], axis=0)
d.columns = ["STATUS", "Symbol"]
d.to_csv("RES_initial_dataload.csv", index=False)

d

STATUS Symbol
0     OK   AAPL
1     OK   TWTR
2     OK   XXII

In [13]:
ticker_list = rr.iloc[0:100]["Symbol"].values
#ticker_list = rr["Symbol"].values
_r = initial_dataload(ticker_list, verbose=True, del_temp=True)
ticker_list

(1/100) Recv. and Stored XXII
(2/100) Recv. and Stored FAX
(3/100) Recv. and Stored IAF
(4/100) Recv. and Stored CH
(5/100) ERROR receiving ABE           
(6/100) Recv. and Stored FCO
(7/100) Recv. and Stored IF
(8/100) Recv. and Stored ISL
(9/100) Recv. and Stored ACU
(10/100) Recv. and Stored AIII
(11/100) Recv. and Stored ATNM
(12/100) Recv. and Stored AE
(13/100) Recv. and Stored ADK
(14/100) ERROR receiving ADK^A
(15/100) Recv. and Stored ACY
(16/100) Recv. and Stored WGA
(17/100) Recv. and Stored AIRI
(18/100) ERROR receiving AA^
(19/100) ERROR receiving AXX
(20/100) Recv. and Stored AXU
(21/100) Recv. and Stored AAU
(22/100) Recv. and Stored APT
(23/100) ERROR receiving ALTV
(24/100) Recv. and Stored AAMC
(25/100) Recv. and Stored DIT
(26/100) ERROR receiving APP
(27/100) Recv. and Stored ADGE
(28/100) Recv. and Stored ALN
(29/100) Recv. and Stored AMS
(30/100) Recv. and Stored AMPE
(31/100) Recv. and Stored AXN
(32/100) Recv. and Stored AMCO
(33/100) Recv. and Stored HRT
(34/10

array(['XXII', 'FAX', 'IAF', 'CH', 'ABE           ', 'FCO', 'IF', 'ISL',
       'ACU', 'AIII', 'ATNM', 'AE', 'ADK', 'ADK^A', 'ACY', 'WGA', 'AIRI',
       'AA^', 'AXX', 'AXU', 'AAU', 'APT', 'ALTV', 'AAMC', 'DIT', 'APP',
       'ADGE', 'ALN', 'AMS', 'AMPE', 'AXN', 'AMCO', 'HRT', 'AKG', 'AINC',
       'ASB.WS', 'AST', 'AWX', 'AVL', 'ASM', 'BTG', 'BTN', 'BKJ', 'BCV',
       'BAA', 'BHB', 'BRN', 'BIR^A', 'BGSF', 'BPMX', 'BTX', 'BTX.WS',
       'BGI', 'BZM', 'MHE', 'BLE', 'BLJ', 'BFY', 'BHV', 'BDR', 'BRG',
       'BVX', 'BWL.A', 'BZC', 'BTI', 'CANF', 'ROX', 'CAW', 'CVM', 'CVM.WS',
       'CEF', 'GTU', 'CET', 'LEU', 'CCF', 'CQH', 'CQP', 'LNG', 'CVR',
       'CNR', 'CPHI', 'CKX', 'GLV', 'GLQ', 'GLO', 'CRV', 'MOC', 'CIX',
       'LODE', 'CTO', 'MCF', 'CUO', 'CMT', 'CVRS', 'CRMD', 'CRF', 'CLM',
       'CVU', 'CIK', 'DHY'], dtype=object)

^ It is nicer to have the time from start to finish of this

In [ ]:
start = datetime.datetime.now()

In [ ]:
end = datetime.datetime.now()

In [ ]:
d = end-start

In [ ]:
d.seconds

In [ ]:
print(str(d))

> Let's try to read back in one of the files we just created....

In [ ]:
XXII = pd.read_csv("{}/{}.csv".format(DATA_PATH, "XXII"))

print("Shape: {}".format(XXII.shape))
print("\nStats\n")
print(XXII.describe())

> Lets see what happens if the file does not exist....
> We might want to consider the FileNotFoundError...

In [ ]:
try:
    XXIIAAAAAAAAA = pd.read_csv("{}/{}.csv".format(DATA_PATH, "XXIIAAAAAAAAA"))
except (FileNotFoundError):
    print("File not found.")
    print("We could create a function to download the data instead of erroring...")

> While we're here, let's look at the diference between csv and hdfs...

In [73]:
_tkrs = ["AAPL", "TWTR", "XXII", "AMZN", "GOOG"]

_r = initial_dataload(_tkrs)

for t in _tkrs:
    itr = pd.read_csv("{}/{}.csv".format(DATA_PATH, t))
    itr.to_hdf("{}/{}.h5".format(DATA_PATH, t), "stock_daily")

(1/5) Recv. and Stored AAPL
(2/5) Recv. and Stored TWTR
(3/5) Recv. and Stored XXII
(4/5) Recv. and Stored AMZN
(5/5) Recv. and Stored GOOG
Took 0:00:01.538005


In [74]:
%%bash
cd data
pwd
ls -lh
du -h

/home/jovyan/work/FintechCapstone/data
total 12M
-rw-r--r-- 1 jovyan users 620K Mar 15 14:51 AAPL.csv
-rw-r--r-- 1 jovyan users 3.9M Mar 15 14:51 AAPL.h5
-rw-r--r-- 1 jovyan users 342K Mar 15 14:51 AMZN.csv
-rw-r--r-- 1 jovyan users 2.0M Mar 15 14:51 AMZN.h5
drwxr-xr-x 2 jovyan users 4.0K Feb 17 14:58 B_TIER1
-rw-r--r-- 1 jovyan users 256K Mar 15 14:51 GOOG.csv
-rw-r--r-- 1 jovyan users 1.7M Mar 15 14:51 GOOG.h5
-rw-r--r-- 1 jovyan users  58K Mar 15 14:51 TWTR.csv
-rw-r--r-- 1 jovyan users 1.3M Mar 15 14:51 TWTR.h5
-rw-r--r-- 1 jovyan users  66K Mar 15 14:51 XXII.csv
-rw-r--r-- 1 jovyan users 1.6M Mar 15 14:51 XXII.h5
14M	./B_TIER1
26M	.


> Ok, so hdf5 will help us manage really large files. But is a bit of overkill for this kind of problem right now

In [75]:
_stock = dict()

_stock["AAPL"] = pd.read_csv("{}/{}.csv".format(DATA_PATH, "AAPL"))
_stock["TWTR"] = pd.read_csv("{}/{}.csv".format(DATA_PATH, "TWTR"))
_stock["XXII"] = pd.read_csv("{}/{}.csv".format(DATA_PATH, "XXII"))
_stock["GOOG"] = pd.read_csv("{}/{}.csv".format(DATA_PATH, "GOOG"))
_stock["AMZN"] = pd.read_csv("{}/{}.csv".format(DATA_PATH, "AMZN"))

> Now, lets settle on a little ticker list

In [76]:
_tkrs_df = pd.DataFrame(_tkrs, columns=["ticker"])
_tkrs_df

ticker
0   AAPL
1   TWTR
2   XXII
3   AMZN
4   GOOG

> Let us take a look at their min and max dates

In [77]:
_tkrs_df["min"] = _tkrs_df["ticker"].apply(lambda s: _stock[s]["Date"].min())
_tkrs_df["max"] = _tkrs_df["ticker"].apply(lambda s: _stock[s]["Date"].max())

_tkrs_df

ticker         min         max
0   AAPL  1980-12-12  2017-03-14
1   TWTR  2013-11-07  2017-03-14
2   XXII  2011-01-26  2017-03-14
3   AMZN  1997-05-15  2017-03-14
4   GOOG  2004-08-19  2017-03-14

> We have been ignoring the fact that the dates come as strings from the csv

In [78]:
_stock["GOOG"]["Date"].head()

0    2004-08-19
1    2004-08-20
2    2004-08-23
3    2004-08-24
4    2004-08-25
Name: Date, dtype: object

> Notice the **dtype** is "object" instead of "date".

> We can take care of this using **to_datetime**

In [79]:
_stock["GOOG"].iloc[3142:3200]["Date"]

3142    2017-02-10
3143    2017-02-13
3144    2017-02-14
3145    2017-02-15
3146    2017-02-16
3147    2017-02-17
3148    2017-02-21
3149    2017-02-22
3150    2017-02-23
3151    2017-02-24
3152    2017-02-27
3153    2017-02-28
3154    2017-03-01
3155    2017-03-02
3156    2017-03-03
3157    2017-03-06
3158    2017-03-07
3159    2017-03-08
3160    2017-03-09
3161    2017-03-10
3162    2017-03-13
3163    2017-03-14
Name: Date, dtype: object

In [80]:
pd.to_datetime(_stock["GOOG"].iloc[:3142]["Date"], infer_datetime_format=True).head()

0   2004-08-19
1   2004-08-20
2   2004-08-23
3   2004-08-24
4   2004-08-25
Name: Date, dtype: datetime64[ns]

In [81]:
def convert_dates(s):
    _stock[s]["Date"] = pd.to_datetime(_stock[s]["Date"], infer_datetime_format=True)

_a = _tkrs_df["ticker"].apply(convert_dates)

print(_stock["GOOG"]["Date"].head())
print("\n")
print(_stock["AAPL"]["Date"].head())

0   2004-08-19
1   2004-08-20
2   2004-08-23
3   2004-08-24
4   2004-08-25
Name: Date, dtype: datetime64[ns]


0   1980-12-12
1   1980-12-15
2   1980-12-16
3   1980-12-17
4   1980-12-18
Name: Date, dtype: datetime64[ns]


In [82]:
_stock["GOOG"].head()

Date        Open        High         Low       Close    Volume  \
0 2004-08-19  100.000168  104.060182   95.960165  100.340176  44871300   
1 2004-08-20  101.010175  109.080187  100.500174  108.310183  22942800   
2 2004-08-23  110.750191  113.480193  109.050183  109.400185  18342800   
3 2004-08-24  111.240189  111.600192  103.570177  104.870176  15319700   
4 2004-08-25  104.960181  108.000187  103.880180  106.000184   9232100   

   Adj Close  
0  50.119968  
1  54.100990  
2  54.645447  
3  52.382705  
4  52.947145

In [83]:
_tkrs_df

ticker         min         max
0   AAPL  1980-12-12  2017-03-14
1   TWTR  2013-11-07  2017-03-14
2   XXII  2011-01-26  2017-03-14
3   AMZN  1997-05-15  2017-03-14
4   GOOG  2004-08-19  2017-03-14

> From now on it is safe to slice the *Date* columns using datetime objects

In [84]:
_stock["GOOG"]["Date"].head() > datetime.datetime(2004, 8, 20)

0    False
1    False
2     True
3     True
4     True
Name: Date, dtype: bool

In [85]:
_stock["GOOG"]["Date"].head() > datetime.datetime(2004, 8, 24)

0    False
1    False
2    False
3    False
4     True
Name: Date, dtype: bool

> Lets take a look at rolling indicators

In [86]:
pd.rolling_mean(_stock["GOOG"]["Open"], 5).iloc[0:14]

/opt/conda/lib/python3.5/site-packages/ipykernel/__main__.py:1: FutureWarning: pd.rolling_mean is deprecated for Series and will be removed in a future version, replace with 
	Series.rolling(center=False,window=5).mean()
  if __name__ == '__main__':


0            NaN
1            NaN
2            NaN
3            NaN
4     105.592181
5     106.582183
6     108.000185
7     106.906183
8     105.118179
9     104.666178
10    103.514177
11    102.084174
12    101.230174
13    100.918174
Name: Open, dtype: float64

In [87]:
_stock["GOOG"]["Open"].rolling(window=5, center=False).mean().iloc[0:14]

0            NaN
1            NaN
2            NaN
3            NaN
4     105.592181
5     106.582183
6     108.000185
7     106.906183
8     105.118179
9     104.666178
10    103.514177
11    102.084174
12    101.230174
13    100.918174
Name: Open, dtype: float64

In [88]:
_stock["GOOG"]["Close"].rolling(window=5, center=False).mean().iloc[0:14]

0            NaN
1            NaN
2            NaN
3            NaN
4     105.784181
5     107.298182
6     106.866182
7     105.388179
8     104.888179
9     103.738176
10    102.458174
11    101.230172
12    101.144173
13    101.130172
Name: Close, dtype: float64

> One can slice a subset of columns using loc and then call apply across all the sliced columns.

> The result will be a new dataframe with the rolled averages row-by-row across each column

In [89]:
def apply_rolling_mean(s):
    return s.rolling(window=5, center=False).mean()

_rolling5 = _stock["GOOG"].loc[:, ["Open", "High", "Low", "Close", "Volume", "Adj Close"]].apply(apply_rolling_mean)
_rolling5.iloc[0:14]


Open        High         Low       Close      Volume  Adj Close
0          NaN         NaN         NaN         NaN         NaN        NaN
1          NaN         NaN         NaN         NaN         NaN        NaN
2          NaN         NaN         NaN         NaN         NaN        NaN
3          NaN         NaN         NaN         NaN         NaN        NaN
4   105.592181  109.244188  102.592176  105.784181  22141740.0  52.839251
5   106.582183  110.022189  104.332179  107.298182  14593200.0  53.595495
6   108.000185  109.930189  105.370180  106.866182  11252880.0  53.379711
7   106.906183  108.332187  103.962178  105.388179   8628600.0  52.641448
8   105.118179  106.754185  103.680178  104.888179   6552900.0  52.391698
9   104.666178  105.748184  102.838175  103.738176   6542800.0  51.817271
10  103.514177  104.632181  101.694174  102.458174   8155160.0  51.177909
11  102.084174  103.256179  100.420171  101.230172   7942280.0  50.564522
12  101.230174  102.558177   99.940171  101.144173   8073040.0  50.521565
13  100.918174  102.422177   99.608170  101.130172   8086640.0  50.514572

> We can concatenate the new frame and the old one using concat (we need to set the *axis=* in order to do it across columns)

> We can also reset the column names so that we don't have duplicate names (we could have done this in the previous step on the _rolling5 frame)

In [90]:
_result = pd.concat([_stock["GOOG"], _rolling5], axis=1)
_result.columns = ['Date', 'Open', 'High', 'Low', 'Close', 'Volume', 'Adj Close', 'Open_roll5','High_roll5', 'Low_roll5', 'Close_roll5', 'Volume_roll5', 'Adj Close_roll5']

_result.iloc[0:14]

Date        Open        High         Low       Close    Volume  \
0  2004-08-19  100.000168  104.060182   95.960165  100.340176  44871300   
1  2004-08-20  101.010175  109.080187  100.500174  108.310183  22942800   
2  2004-08-23  110.750191  113.480193  109.050183  109.400185  18342800   
3  2004-08-24  111.240189  111.600192  103.570177  104.870176  15319700   
4  2004-08-25  104.960181  108.000187  103.880180  106.000184   9232100   
5  2004-08-26  104.950180  107.950188  104.660179  107.910182   7128600   
6  2004-08-27  108.100185  108.620186  105.690180  106.150181   6241200   
7  2004-08-30  105.280178  105.490184  102.010172  102.010172   5221400   
8  2004-08-31  102.300173  103.710180  102.160177  102.370175   4941200   
9  2004-09-01  102.700174  102.970180   99.670169  100.250171   9181600   
10 2004-09-02   99.190173  102.370175   98.940170  101.510173  15190400   
11 2004-09-03  100.950174  101.740174   99.320168  100.010169   5176800   
12 2004-09-07  101.010175  102.000178   99.610169  101.580175   5875200   
13 2004-09-08  100.740176  103.030180  100.500174  102.300173   5009200   

    Adj Close  Open_roll5  High_roll5   Low_roll5  Close_roll5  Volume_roll5  \
0   50.119968         NaN         NaN         NaN          NaN           NaN   
1   54.100990         NaN         NaN         NaN          NaN           NaN   
2   54.645447         NaN         NaN         NaN          NaN           NaN   
3   52.382705         NaN         NaN         NaN          NaN           NaN   
4   52.947145  105.592181  109.244188  102.592176   105.784181    22141740.0   
5   53.901190  106.582183  110.022189  104.332179   107.298182    14593200.0   
6   53.022069  108.000185  109.930189  105.370180   106.866182    11252880.0   
7   50.954132  106.906183  108.332187  103.962178   105.388179     8628600.0   
8   51.133953  105.118179  106.754185  103.680178   104.888179     6552900.0   
9   50.075011  104.666178  105.748184  102.838175   103.738176     6542800.0   
10  50.704382  103.514177  104.632181  101.694174   102.458174     8155160.0   
11  49.955130  102.084174  103.256179  100.420171   101.230172     7942280.0   
12  50.739348  101.230174  102.558177   99.940171   101.144173     8073040.0   
13  51.098988  100.918174  102.422177   99.608170   101.130172     8086640.0   

    Adj Close_roll5  
0               NaN  
1               NaN  
2               NaN  
3               NaN  
4         52.839251  
5         53.595495  
6         53.379711  
7         52.641448  
8         52.391698  
9         51.817271  
10        51.177909  
11        50.564522  
12        50.521565  
13        50.514572

> Lets see how we could repurpose **apply_rolling_mean** to use with different window sizes with an apply over columns...

In [91]:
def apply_rolling_mean(s, window=5):
    return s.rolling(window=window, center=False).mean()

_col_slice = ["Open", "High", "Low", "Close", "Volume", "Adj Close"]
_pre_bound = partial(apply_rolling_mean, window=3)
_rolling = _stock["GOOG"].loc[ : , _col_slice ].apply(_pre_bound)
_rolling.iloc[0:8]


Open        High         Low       Close        Volume  Adj Close
0         NaN         NaN         NaN         NaN           NaN        NaN
1         NaN         NaN         NaN         NaN           NaN        NaN
2  103.920178  108.873521  101.836841  106.016848  2.871897e+07  52.955468
3  107.666852  111.386857  104.373511  107.526848  1.886843e+07  53.709714
4  108.983520  111.026857  105.500180  106.756848  1.429820e+07  53.325099
5  107.050183  109.183522  104.036845  106.260181  1.056013e+07  53.077013
6  106.003515  108.190187  104.743513  106.686849  7.533967e+06  53.290135
7  106.110181  107.353519  104.120177  105.356845  6.197067e+06  52.625797

In [92]:
# We can also call rolling directly from the dataframe itself
_stock["GOOG"].loc[:,["Open"]].rolling(window=3, center=False).mean().head()

Open
0         NaN
1         NaN
2  103.920178
3  107.666852
4  108.983520

In [93]:
# this will be the same as the cells above
a = _stock["GOOG"].loc[:,["Open"]].rolling(window=3, center=False)
getattr(a, 'mean')().head()

Open
0         NaN
1         NaN
2  103.920178
3  107.666852
4  108.983520

In [94]:
# but now we can call other rolling stats programatically

stat_fns = ["mean", "sum"]

for itr in stat_fns:
    print("{} :\n".format(itr))
    print(getattr(a, itr)().head())
    print("\n---\n")

mean :

         Open
0         NaN
1         NaN
2  103.920178
3  107.666852
4  108.983520

---

sum :

         Open
0         NaN
1         NaN
2  311.760534
3  323.000555
4  326.950561

---



> We can bundle all this rolling logic in a function that rolls and merges the collumns that we want

In [95]:
roll_columns(_stock["GOOG"], "mean", column_slice=["Open"]).iloc[0:16]

Open_roll_mean_10
0                 NaN
1                 NaN
2                 NaN
3                 NaN
4                 NaN
5                 NaN
6                 NaN
7                 NaN
8                 NaN
9          105.129179
10         105.048180
11         105.042180
12         104.068178
13         103.018177
14         102.775176
15         102.440176

In [96]:
roll_columns(_stock["GOOG"], "mean", column_slice=["Open", "Close"]).iloc[0:16]

Open_roll_mean_10  Close_roll_mean_10
0                 NaN                 NaN
1                 NaN                 NaN
2                 NaN                 NaN
3                 NaN                 NaN
4                 NaN                 NaN
5                 NaN                 NaN
6                 NaN                 NaN
7                 NaN                 NaN
8                 NaN                 NaN
9          105.129179          104.761178
10         105.048180          104.878178
11         105.042180          104.048177
12         104.068178          103.266176
13         103.018177          103.009175
14         102.775176          102.640174
15         102.440176          102.382174

In [97]:
roll_columns(_stock["GOOG"], "mean", column_slice=["Open", "Close"], window=5).head()

Open_roll_mean_5  Close_roll_mean_5
0               NaN                NaN
1               NaN                NaN
2               NaN                NaN
3               NaN                NaN
4        105.592181         105.784181

In [98]:
roll_columns(_stock["GOOG"], "mean", column_slice=["Open", "Close"], window=3).head()

Open_roll_mean_3  Close_roll_mean_3
0               NaN                NaN
1               NaN                NaN
2        103.920178         106.016848
3        107.666852         107.526848
4        108.983520         106.756848

In [99]:
roll_columns(_stock["GOOG"], "mean", column_slice=["Open", "Close"], window=3, merge_result=True).head()

Date        Open        High         Low       Close    Volume  \
0 2004-08-19  100.000168  104.060182   95.960165  100.340176  44871300   
1 2004-08-20  101.010175  109.080187  100.500174  108.310183  22942800   
2 2004-08-23  110.750191  113.480193  109.050183  109.400185  18342800   
3 2004-08-24  111.240189  111.600192  103.570177  104.870176  15319700   
4 2004-08-25  104.960181  108.000187  103.880180  106.000184   9232100   

   Adj Close  Open_roll_mean_3  Close_roll_mean_3  
0  50.119968               NaN                NaN  
1  54.100990               NaN                NaN  
2  54.645447        103.920178         106.016848  
3  52.382705        107.666852         107.526848  
4  52.947145        108.983520         106.756848

> We really shouldn't leave the nan or other missing values there.

In [100]:
df_topad = roll_columns(_stock["GOOG"], "mean", column_slice=["Open", "Close"], window=3).head()
df_topad

Open_roll_mean_3  Close_roll_mean_3
0               NaN                NaN
1               NaN                NaN
2        103.920178         106.016848
3        107.666852         107.526848
4        108.983520         106.756848

In [101]:
df_topad.fillna(method="bfill")

Open_roll_mean_3  Close_roll_mean_3
0        103.920178         106.016848
1        103.920178         106.016848
2        103.920178         106.016848
3        107.666852         107.526848
4        108.983520         106.756848

> we could acomodate this on our function....

In [102]:
roll_columns(_stock["GOOG"], "mean", column_slice=["Open", "Close"], window=3, pad_result=False).head()

Open_roll_mean_3  Close_roll_mean_3
0               NaN                NaN
1               NaN                NaN
2        103.920178         106.016848
3        107.666852         107.526848
4        108.983520         106.756848

In [103]:
roll_columns(_stock["GOOG"], "mean", column_slice=["Open", "Close"], window=3, pad_result=True).head()

Open_roll_mean_3  Close_roll_mean_3
0        103.920178         106.016848
1        103.920178         106.016848
2        103.920178         106.016848
3        107.666852         107.526848
4        108.983520         106.756848

We can look at how to compute the daily returns by shifting our data

In [104]:
_sample_goog_close = _stock["GOOG"].loc[:,["Close"]].iloc[:15]
_sample_goog_close

Close
0   100.340176
1   108.310183
2   109.400185
3   104.870176
4   106.000184
5   107.910182
6   106.150181
7   102.010172
8   102.370175
9   100.250171
10  101.510173
11  100.010169
12  101.580175
13  102.300173
14  102.310174

In [105]:
%%timeit

_rr = (_sample_goog_close[1:] / _sample_goog_close[:-1].values) - 1
_rr.ix[0] = 0
_rr



100 loops, best of 3: 4.05 ms per loop


In [106]:
%%timeit

# this code does the same as the cell above, but as we can see it is much faster
_rr = (_sample_goog_close / _sample_goog_close.shift(1)) - 1
_rr.ix[0] = 0
_rr

1000 loops, best of 3: 1.14 ms per loop


> Lets try with multiple collumns

In [107]:

_rr = (_sample_goog_close / _sample_goog_close.shift(1)) - 1
_rr.ix[0] = 0
_rr

Close
0   0.000000
1   0.079430
2   0.010064
3  -0.041408
4   0.010775
5   0.018019
6  -0.016310
7  -0.039001
8   0.003529
9  -0.020709
10  0.012569
11 -0.014777
12  0.015698
13  0.007088
14  0.000098

In [108]:
_rr = (_sample_goog_close / _sample_goog_close.shift(1)) - 1
_rr = _rr.fillna(0)
_rr

Close
0   0.000000
1   0.079430
2   0.010064
3  -0.041408
4   0.010775
5   0.018019
6  -0.016310
7  -0.039001
8   0.003529
9  -0.020709
10  0.012569
11 -0.014777
12  0.015698
13  0.007088
14  0.000098

> Lets try out some functions of our own. Built on top of this

In [109]:
import numpy as np
df = pd.DataFrame([[1, 1.2, 1.8],[1, 0.8, 1.6],[1,-0.5,0.1]]).T
df

0    1    2
0  1.0  1.0  1.0
1  1.2  0.8 -0.5
2  1.8  1.6  0.1

In [110]:
timewindow_diff(df)

timewindow_diff_1_0  timewindow_diff_1_1  timewindow_diff_1_2
0                  NaN                  NaN                  NaN
1                  0.2                 -0.2                 -1.5
2                  0.6                  0.8                  0.6

In [111]:
timewindow_diff(df, shift_window=2, fillna=True)

timewindow_diff_2_0  timewindow_diff_2_1  timewindow_diff_2_2
0                  0.0                  0.0                  0.0
1                  0.0                  0.0                  0.0
2                  0.8                  0.6                 -0.9

In [112]:
timewindow_diff(df, column_slice=[1], fillna=True, merge_result=True)

0    1    2  timewindow_diff_1_1
0  1.0  1.0  1.0                  0.0
1  1.2  0.8 -0.5                 -0.2
2  1.8  1.6  0.1                  0.8

In [113]:
df

0    1    2
0  1.0  1.0  1.0
1  1.2  0.8 -0.5
2  1.8  1.6  0.1

In [114]:
timewindow_return(df)

timewindow_return_1_0  timewindow_return_1_1  timewindow_return_1_2
0                    NaN                    NaN                    NaN
1                    0.2                   -0.2                   -1.5
2                    0.5                    1.0                   -1.2

> timewindow_return gives us the variation from an initial value. See the explanation below.

In [115]:
n = 200
p = 1.5
v = p * n
print("You buy {} shares at {}. Your value is {}".format(n,p,v))
np = 1.0
nv = np * n
r = ((np / p) - 1.0)
print("The price changes to {}. You new value is {}.".format(np, nv))

print ("Your return is {} ({} of your initial value)".format(v*r, r))

You buy 200 shares at 1.5. Your value is 300.0
The price changes to 1.0. You new value is 200.0.
Your return is -100.00000000000001 (-0.33333333333333337 of your initial value)


In [116]:
df

0    1    2
0  1.0  1.0  1.0
1  1.2  0.8 -0.5
2  1.8  1.6  0.1

In [117]:
timewindow_cumdiff(df)

timewindow_cumdiff_0  timewindow_cumdiff_1  timewindow_cumdiff_2
0                   0.0                   0.0                   0.0
1                   0.2                  -0.2                  -1.5
2                   0.8                   0.6                  -0.9

In [118]:
timewindow_cumreturn(df)

timewindow_cumreturn_0  timewindow_cumreturn_1  timewindow_cumreturn_2
0                     0.0                     0.0                     0.0
1                     0.2                    -0.2                    -1.5
2                     0.8                     0.6                    -0.9

> Now, lets take a real look at dates and how to manage them

In [119]:
import numpy as np

# This can create an array with a date for us
ledate = np.array('2015-12-30', dtype=np.datetime64)
ledate

array(datetime.date(2015, 12, 30), dtype='datetime64[D]')

> we can obviously do vectorized operation on *ledate*

In [120]:
lerange_12day = ledate + np.arange(12)
lerange_12day

array(['2015-12-30', '2015-12-31', '2016-01-01', '2016-01-02',
       '2016-01-03', '2016-01-04', '2016-01-05', '2016-01-06',
       '2016-01-07', '2016-01-08', '2016-01-09', '2016-01-10'], dtype='datetime64[D]')

> We can also use *np.datetime64* in order to create a date constant. It's possible to give it a time format

In [121]:
print(np.datetime64('2015-07-04 12:00'))
print(np.datetime64('2015-07-04 12:59:59.50', 'ns'))

2015-07-04T12:00
2015-07-04T12:59:59.500000000


> These are the code formats possible with numpy

|Code    | Meaning     | Time span (relative) | Time span (absolute)   |
|--------|-------------|----------------------|------------------------|
| ``Y``  | Year	       | ± 9.2e18 years       | [9.2e18 BC, 9.2e18 AD] |
| ``M``  | Month       | ± 7.6e17 years       | [7.6e17 BC, 7.6e17 AD] |
| ``W``  | Week	       | ± 1.7e17 years       | [1.7e17 BC, 1.7e17 AD] |
| ``D``  | Day         | ± 2.5e16 years       | [2.5e16 BC, 2.5e16 AD] |
| ``h``  | Hour        | ± 1.0e15 years       | [1.0e15 BC, 1.0e15 AD] |
| ``m``  | Minute      | ± 1.7e13 years       | [1.7e13 BC, 1.7e13 AD] |
| ``s``  | Second      | ± 2.9e12 years       | [ 2.9e9 BC, 2.9e9 AD]  |
| ``ms`` | Millisecond | ± 2.9e9 years        | [ 2.9e6 BC, 2.9e6 AD]  |
| ``us`` | Microsecond | ± 2.9e6 years        | [290301 BC, 294241 AD] |
| ``ns`` | Nanosecond  | ± 292 years          | [ 1678 AD, 2262 AD]    |
| ``ps`` | Picosecond  | ± 106 days           | [ 1969 AD, 1970 AD]    |
| ``fs`` | Femtosecond | ± 2.6 hours          | [ 1969 AD, 1970 AD]    |
| ``as`` | Attosecond  | ± 9.2 seconds        | [ 1969 AD, 1970 AD]    |

> We can also use pandas to manipulate our dates

In [122]:
ledate = pd.to_datetime("4th of July, 2015")
ledate

Timestamp('2015-07-04 00:00:00')

In [123]:
ledate.strftime('%A')

'Saturday'

> We can also use vectorized operations 

In [124]:
lerange_12days = ledate + pd.to_timedelta(np.arange(12), 'D')
lerange_12days

DatetimeIndex(['2015-07-04', '2015-07-05', '2015-07-06', '2015-07-07',
               '2015-07-08', '2015-07-09', '2015-07-10', '2015-07-11',
               '2015-07-12', '2015-07-13', '2015-07-14', '2015-07-15'],
              dtype='datetime64[ns]', freq=None)

In [125]:
lerange_12months = ledate + pd.to_timedelta(np.arange(12), 'M')
lerange_12months

DatetimeIndex(['2015-07-04 00:00:00', '2015-08-03 10:29:06',
               '2015-09-02 20:58:12', '2015-10-03 07:27:18',
               '2015-11-02 17:56:24', '2015-12-03 04:25:30',
               '2016-01-02 14:54:36', '2016-02-02 01:23:42',
               '2016-03-03 11:52:48', '2016-04-02 22:21:54',
               '2016-05-03 08:51:00', '2016-06-02 19:20:06'],
              dtype='datetime64[ns]', freq=None)

In [126]:
lerange_12minutes = ledate + pd.to_timedelta(np.arange(12), 'm')
lerange_12minutes

DatetimeIndex(['2015-07-04 00:00:00', '2015-07-04 00:01:00',
               '2015-07-04 00:02:00', '2015-07-04 00:03:00',
               '2015-07-04 00:04:00', '2015-07-04 00:05:00',
               '2015-07-04 00:06:00', '2015-07-04 00:07:00',
               '2015-07-04 00:08:00', '2015-07-04 00:09:00',
               '2015-07-04 00:10:00', '2015-07-04 00:11:00'],
              dtype='datetime64[ns]', freq=None)

> We can also use *pd.date_range* to achieve this

In [127]:
pd.date_range('2015-07-03', '2015-07-10')

DatetimeIndex(['2015-07-03', '2015-07-04', '2015-07-05', '2015-07-06',
               '2015-07-07', '2015-07-08', '2015-07-09', '2015-07-10'],
              dtype='datetime64[ns]', freq='D')

In [128]:
pd.date_range('2015-07-03', periods=4)

DatetimeIndex(['2015-07-03', '2015-07-04', '2015-07-05', '2015-07-06'], dtype='datetime64[ns]', freq='D')

In [129]:
pd.date_range('2015-07-03', periods=8, freq='H')

DatetimeIndex(['2015-07-03 00:00:00', '2015-07-03 01:00:00',
               '2015-07-03 02:00:00', '2015-07-03 03:00:00',
               '2015-07-03 04:00:00', '2015-07-03 05:00:00',
               '2015-07-03 06:00:00', '2015-07-03 07:00:00'],
              dtype='datetime64[ns]', freq='H')

In [130]:
pd.period_range('2015-07', periods=8, freq='M')

PeriodIndex(['2015-07', '2015-08', '2015-09', '2015-10', '2015-11', '2015-12',
             '2016-01', '2016-02'],
            dtype='period[M]', freq='M')

In [131]:
pd.timedelta_range(0, periods=10, freq='H')

TimedeltaIndex(['00:00:00', '01:00:00', '02:00:00', '03:00:00', '04:00:00',
                '05:00:00', '06:00:00', '07:00:00', '08:00:00', '09:00:00'],
               dtype='timedelta64[ns]', freq='H')

#### Frequencies and Offsets
(Taken from [Python Data Science Handbook](http://shop.oreilly.com/product/0636920034919.do) by Jake VanderPlas; the content is available [on GitHub](https://github.com/jakevdp/PythonDataScienceHandbook))

Fundamental to these Pandas time series tools is the concept of a frequency or date offset.
Just as we saw the ``D`` (day) and ``H`` (hour) codes above, we can use such codes to specify any desired frequency spacing.
The following table summarizes the main codes available:

| Code   | Description         | Code   | Description          |
|--------|---------------------|--------|----------------------|
| ``D``  | Calendar day        | ``B``  | Business day         |
| ``W``  | Weekly              |        |                      |
| ``M``  | Month end           | ``BM`` | Business month end   |
| ``Q``  | Quarter end         | ``BQ`` | Business quarter end |
| ``A``  | Year end            | ``BA`` | Business year end    |
| ``H``  | Hours               | ``BH`` | Business hours       |
| ``T``  | Minutes             |        |                      |
| ``S``  | Seconds             |        |                      |
| ``L``  | Milliseonds         |        |                      |
| ``U``  | Microseconds        |        |                      |
| ``N``  | nanoseconds         |        |                      |

| Code    | Description            | Code    | Description            |
|---------|------------------------|---------|------------------------|
| ``MS``  | Month start            |``BMS``  | Business month start   |
| ``QS``  | Quarter start          |``BQS``  | Business quarter start |
| ``AS``  | Year start             |``BAS``  | Business year start    |

In [132]:
## This will give us 9 spans of 2 and a half hours each
pd.timedelta_range(0, periods=9, freq="2H30T")

TimedeltaIndex(['00:00:00', '02:30:00', '05:00:00', '07:30:00', '10:00:00',
                '12:30:00', '15:00:00', '17:30:00', '20:00:00'],
               dtype='timedelta64[ns]', freq='150T')

In [133]:
aapl = _stock["AAPL"]
aapl.head()

Date       Open       High     Low   Close     Volume  Adj Close
0 1980-12-12  28.750000  28.875000  28.750  28.750  117258400   0.425000
1 1980-12-15  27.375001  27.375001  27.250  27.250   43971200   0.402826
2 1980-12-16  25.375000  25.375000  25.250  25.250   26432000   0.373261
3 1980-12-17  25.875000  25.999999  25.875  25.875   21610400   0.382500
4 1980-12-18  26.625000  26.750000  26.625  26.625   18362400   0.393587

In [134]:
aapl.corr()

Open      High       Low     Close    Volume  Adj Close
Open       1.000000  0.999929  0.999902  0.999826  0.200913   0.603955
High       0.999929  1.000000  0.999882  0.999921  0.203637   0.603304
Low        0.999902  0.999882  1.000000  0.999915  0.196851   0.604890
Close      0.999826  0.999921  0.999915  1.000000  0.200406   0.604161
Volume     0.200913  0.203637  0.196851  0.200406  1.000000   0.013555
Adj Close  0.603955  0.603304  0.604890  0.604161  0.013555   1.000000

In [135]:
goog = _stock["GOOG"]

In [136]:
_rolling = aapl.loc[:100, ["Close"]].rolling(window=10, center=True)

_rolling.corr(other=goog["Open"]).iloc[:16]

Close
0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
5  -0.492011
6  -0.696666
7  -0.874905
8  -0.850180
9  -0.807468
10 -0.850498
11 -0.855884
12 -0.811677
13 -0.844038
14 -0.784980
15 -0.783991

In [137]:
aapl["Close"].corr(goog["Close"])

0.49542591210516851

> Lets take a look at the exchanges

In [138]:
_tkrs = ["^IXIC", "^GSPC"]

_r = initial_dataload(_tkrs)

_exchanges = dict()

for t in _tkrs:
    _exchanges[t] = pd.read_csv("{}/{}.csv".format(DATA_PATH, t))

(1/2) Recv. and Stored ^IXIC
(2/2) Recv. and Stored ^GSPC
Took 0:00:01.651249


In [139]:
df_beta = pd.DataFrame()

df_beta["IXIC"] = _exchanges["^IXIC"].iloc[-100:, :]["Close"]
df_beta["GOOG"] = _stock["GOOG"].iloc[-100:, :]["Close"]

In [140]:
def calc_beta(df):
	# first column is the market
	X = df.values[:, [0]]
	# prepend a column of ones for the intercept
	X = np.concatenate([np.ones_like(X), X], axis=1)
	# matrix algebra
	b = np.linalg.pinv(X.T.dot(X)).dot(X.T).dot(df.values[:, 1:])
	return pd.Series(b[1], df.columns[1:], name='Beta')




In [141]:
df_beta.head()

IXIC  GOOG
11527  5246.410156   NaN
11528  5241.830078   NaN
11529  5257.399902   NaN
11530  5309.830078   NaN
11531  5283.399902   NaN

In [142]:
_stock["GOOG"].iloc[-100:, :]["Close"]

3064    801.500000
3065    796.969971
3066    799.369995
3067    813.109985
3068    807.669983
3069    799.070007
3070    795.349976
3071    795.369995
3072    784.539978
3073    783.609985
3074    768.700012
3075    762.130005
3076    762.020020
3077    782.520020
3078    790.510010
3079    785.309998
3080    762.559998
3081    754.020020
3082    736.080017
3083    758.489990
3084    764.479980
3085    771.229980
3086    760.539978
3087    769.200012
3088    768.270020
3089    760.989990
3090    761.679993
3091    768.239990
3092    770.840027
3093    758.039978
           ...    
3134    796.789978
3135    795.695007
3136    798.530029
3137    801.489990
3138    801.340027
3139    806.969971
3140    808.380005
3141    809.559998
3142    813.669983
3143    819.239990
3144    820.450012
3145    818.979980
3146    824.159973
3147    828.070007
3148    831.659973
3149    830.760010
3150    831.330017
3151    828.640015
3152    829.280029
3153    823.210022
3154    835.239990
3155    830.

In [143]:
dates = pd.date_range('1995-12-31', periods=480, freq='M', name='Date')
stoks = pd.Index(['s{:04d}'.format(i) for i in range(4000)])
df = pd.DataFrame(np.random.rand(480, 4000), dates, stoks)

In [144]:
df.iloc[:5, :5]


s0000     s0001     s0002     s0003     s0004
Date                                                        
1995-12-31  0.439742  0.244586  0.939956  0.793016  0.498755
1996-01-31  0.751395  0.161854  0.319897  0.999553  0.627204
1996-02-29  0.681048  0.976193  0.980843  0.859208  0.268533
1996-03-31  0.684112  0.731804  0.872493  0.846800  0.610690
1996-04-30  0.044781  0.855973  0.631492  0.314742  0.058679

In [145]:
def roll(df, w):
    # stack df.values w-times shifted once at each stack
    roll_array = np.dstack([df.values[i:i+w, :] for i in range(len(df.index) - w + 1)]).T
    # roll_array is now a 3-D array and can be read into
    # a pandas panel object
    panel = pd.Panel(roll_array, 
                     items=df.index[w-1:],
                     major_axis=df.columns,
                     minor_axis=pd.Index(range(w), name='roll'))
    # convert to dataframe and pivot + groupby
    # is now ready for any action normally performed
    # on a groupby object
    return panel.to_frame().unstack().T.groupby(level=0)


In [146]:
def beta(df):
    # first column is the market
    X = df.values[:, [0]]
    # prepend a column of ones for the intercept
    X = np.concatenate([np.ones_like(X), X], axis=1)
    # matrix algebra
    b = np.linalg.pinv(X.T.dot(X)).dot(X.T).dot(df.values[:, 1:])
    return pd.Series(b[1], df.columns[1:], name='Beta')


In [147]:
rdf = roll(df, 12)
betas = rdf.apply(beta)


In [148]:
def calc_beta(df):
    np_array = df.values
    m = np_array[:,0] # market returns are column zero from numpy array
    s = np_array[:,1] # stock returns are column one from numpy array
    covariance = np.cov(s,m) # Calculate covariance between stock and market
    beta = covariance[0,1]/covariance[1,1]
    return beta


In [149]:
print(calc_beta(df.iloc[:12, :2]))


-0.098394398325


In [150]:
print(beta(df.iloc[:12, :2]))


s0001   -0.098394
Name: Beta, dtype: float64


In [151]:
betas = rdf.apply(beta)
betas.iloc[:5, :5]


major          s0001     s0002     s0003     s0004     s0005
Date                                                        
1996-11-30 -0.098394 -0.184887  0.090722 -0.121429  0.550546
1996-12-31 -0.120075 -0.103704  0.140081 -0.142751  0.564142
1997-01-31 -0.132794  0.000154  0.066946 -0.158786  0.473184
1997-02-28 -0.243285 -0.046319  0.077253 -0.071332  0.458667
1997-03-31 -0.268293 -0.094814  0.007051 -0.125453  0.379129

In [152]:
%%timeit 
roll(df, 12).apply(beta)

1 loop, best of 3: 4.88 s per loop


In [153]:
from scipy import stats
import numpy as np
x = np.random.random(10)
y = np.random.random(10)
slope, intercept, r_value, p_value, std_err = stats.linregress(x,y)


In [154]:
_stock = dict()
_stock["TWTR"] = pd.read_csv("{}/{}.csv".format(DATA_PATH, "TWTR"))

In [155]:
timespan = None

timespan = {
    "short_term": [5]
    ,"medium_term": [30,50,70]
    ,"long_term": [100, 200]
}


In [156]:
gains = None

gains = {
    "low": [0.01]
    ,"medium": [0.05, 0.07, 0.1]
    ,"high": [0.2, 0.4, 1.0]
}

In [157]:
losses = None

losses = {
    "low": [-0.05, -0.1]
    ,"medium": [-0.3, -0.5]
    ,"high": [-0.8, -1.0, -2.0]
}

In [158]:
itr_df = _stock["TWTR"];

itr_df.head()

Date       Open       High        Low      Close     Volume  \
0  2013-11-07  45.099998  50.090000  44.000000  44.900002  117701600   
1  2013-11-08  45.930000  46.939999  40.689999  41.650002   27925300   
2  2013-11-11  40.500000  43.000000  39.400002  42.900002   16113900   
3  2013-11-12  43.660000  43.779999  41.830002  41.900002    6316700   
4  2013-11-13  41.029999  42.869999  40.759998  42.599998    8688300   

   Adj Close  
0  44.900002  
1  41.650002  
2  42.900002  
3  41.900002  
4  42.599998

In [159]:
## calculate diff moves for timespans
## calculate returns for timespans
## calculate moving averages
## calculate bollinger bands

In [160]:
## calculate diff moves for timespans

itr_df = _stock["TWTR"];

for ts_name in timespan:
    for t in timespan[ts_name]:
        itr_df = timewindow_diff(itr_df, column_slice=["Close", "Volume"], shift_window=t, merge_result=True)        

itr_df.head()


Date       Open       High        Low      Close     Volume  \
0  2013-11-07  45.099998  50.090000  44.000000  44.900002  117701600   
1  2013-11-08  45.930000  46.939999  40.689999  41.650002   27925300   
2  2013-11-11  40.500000  43.000000  39.400002  42.900002   16113900   
3  2013-11-12  43.660000  43.779999  41.830002  41.900002    6316700   
4  2013-11-13  41.029999  42.869999  40.759998  42.599998    8688300   

   Adj Close  timewindow_diff_5_Close  timewindow_diff_5_Volume  \
0  44.900002                      NaN                       NaN   
1  41.650002                      NaN                       NaN   
2  42.900002                      NaN                       NaN   
3  41.900002                      NaN                       NaN   
4  42.599998                      NaN                       NaN   

   timewindow_diff_30_Close  timewindow_diff_30_Volume  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

   timewindow_diff_50_Close  timewindow_diff_50_Volume  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

   timewindow_diff_70_Close  timewindow_diff_70_Volume  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

   timewindow_diff_100_Close  timewindow_diff_100_Volume  \
0                        NaN                         NaN   
1                        NaN                         NaN   
2                        NaN                         NaN   
3                        NaN                         NaN   
4                        NaN                         NaN   

   timewindow_diff_200_Close  timewindow_diff_200_Volume  
0                        NaN                         NaN  
1                        NaN                         NaN  
2                        NaN                         NaN  
3                        NaN                         NaN  
4                        NaN                         NaN

In [161]:
## calculate returns for timespans

itr_df = _stock["TWTR"];

for ts_name in timespan:
    for t in timespan[ts_name]:
        itr_df = timewindow_return(itr_df, column_slice=["Close", "Volume"], shift_window=t, merge_result=True)        

itr_df.head()



Date       Open       High        Low      Close     Volume  \
0  2013-11-07  45.099998  50.090000  44.000000  44.900002  117701600   
1  2013-11-08  45.930000  46.939999  40.689999  41.650002   27925300   
2  2013-11-11  40.500000  43.000000  39.400002  42.900002   16113900   
3  2013-11-12  43.660000  43.779999  41.830002  41.900002    6316700   
4  2013-11-13  41.029999  42.869999  40.759998  42.599998    8688300   

   Adj Close  timewindow_return_5_Close  timewindow_return_5_Volume  \
0  44.900002                        NaN                         NaN   
1  41.650002                        NaN                         NaN   
2  42.900002                        NaN                         NaN   
3  41.900002                        NaN                         NaN   
4  42.599998                        NaN                         NaN   

   timewindow_return_30_Close  timewindow_return_30_Volume  \
0                         NaN                          NaN   
1                         NaN                          NaN   
2                         NaN                          NaN   
3                         NaN                          NaN   
4                         NaN                          NaN   

   timewindow_return_50_Close  timewindow_return_50_Volume  \
0                         NaN                          NaN   
1                         NaN                          NaN   
2                         NaN                          NaN   
3                         NaN                          NaN   
4                         NaN                          NaN   

   timewindow_return_70_Close  timewindow_return_70_Volume  \
0                         NaN                          NaN   
1                         NaN                          NaN   
2                         NaN                          NaN   
3                         NaN                          NaN   
4                         NaN                          NaN   

   timewindow_return_100_Close  timewindow_return_100_Volume  \
0                          NaN                           NaN   
1                          NaN                           NaN   
2                          NaN                           NaN   
3                          NaN                           NaN   
4                          NaN                           NaN   

   timewindow_return_200_Close  timewindow_return_200_Volume  
0                          NaN                           NaN  
1                          NaN                           NaN  
2                          NaN                           NaN  
3                          NaN                           NaN  
4                          NaN                           NaN

In [162]:
## calculate moving averages

itr_df = _stock["TWTR"];

for ts_name in timespan:
    for t in timespan[ts_name]:
        itr_df = roll_columns(itr_df, "mean", column_slice=["Close", "Volume"], window=t, merge_result=True)        

itr_df.head()


Date       Open       High        Low      Close     Volume  \
0  2013-11-07  45.099998  50.090000  44.000000  44.900002  117701600   
1  2013-11-08  45.930000  46.939999  40.689999  41.650002   27925300   
2  2013-11-11  40.500000  43.000000  39.400002  42.900002   16113900   
3  2013-11-12  43.660000  43.779999  41.830002  41.900002    6316700   
4  2013-11-13  41.029999  42.869999  40.759998  42.599998    8688300   

   Adj Close  Close_roll_mean_5  Volume_roll_mean_5  Close_roll_mean_30  \
0  44.900002                NaN                 NaN                 NaN   
1  41.650002                NaN                 NaN                 NaN   
2  42.900002                NaN                 NaN                 NaN   
3  41.900002                NaN                 NaN                 NaN   
4  42.599998          42.790001          35349160.0                 NaN   

   Volume_roll_mean_30  Close_roll_mean_50  Volume_roll_mean_50  \
0                  NaN                 NaN                  NaN   
1                  NaN                 NaN                  NaN   
2                  NaN                 NaN                  NaN   
3                  NaN                 NaN                  NaN   
4                  NaN                 NaN                  NaN   

   Close_roll_mean_70  Volume_roll_mean_70  Close_roll_mean_100  \
0                 NaN                  NaN                  NaN   
1                 NaN                  NaN                  NaN   
2                 NaN                  NaN                  NaN   
3                 NaN                  NaN                  NaN   
4                 NaN                  NaN                  NaN   

   Volume_roll_mean_100  Close_roll_mean_200  Volume_roll_mean_200  
0                   NaN                  NaN                   NaN  
1                   NaN                  NaN                   NaN  
2                   NaN                  NaN                   NaN  
3                   NaN                  NaN                   NaN  
4                   NaN                  NaN                   NaN

In [163]:
## calculate bollinger bands

itr_df = pd.DataFrame()
res_df = pd.DataFrame()

for ts_name in timespan:
    if ts_name in ["medium_term", "long_term"]:
        for t in timespan[ts_name]:
            itr_df = roll_columns(_stock["TWTR"], "std", column_slice=["Close"], window=t, merge_result=False, scaler=2, pad_result=True)
            upper_band = itr_df.apply(lambda x: s + x)
            lower_band = itr_df.apply(lambda x: s - x)
            
            roll_name = "{}_roll_2std_{}".format("Close", t)
            upper_name = "{}_bollinger_{}_up".format("Close", t)
            lower_name = "{}_bollinger_{}_low".format("Close", t)
            
            res_df[roll_name] = itr_df.iloc[:, 0]
            res_df[upper_name] = upper_band
            res_df[lower_name] = lower_band

            
res_df.head()

NameError: ("name 's' is not defined", 'occurred at index Close_roll_std_30')

In [12]:
_stock = dict()
_stock["TWTR"] = pd.read_csv("{}/{}.csv".format(DATA_PATH, "TWTR"))

In [13]:
aapl = load_raw_frame("AAPL")
aapl.head()

Date       Open       High     Low   Close     Volume  Adj Close
0  1980-12-12  28.750000  28.875000  28.750  28.750  117258400   0.425000
1  1980-12-15  27.375001  27.375001  27.250  27.250   43971200   0.402826
2  1980-12-16  25.375000  25.375000  25.250  25.250   26432000   0.373261
3  1980-12-17  25.875000  25.999999  25.875  25.875   21610400   0.382500
4  1980-12-18  26.625000  26.750000  26.625  26.625   18362400   0.393587

In [14]:
timespan = None

timespan = {
    "medium_term": [4,7]
}


In [15]:
aapl = load_raw_frame("AAPL")

In [16]:
calc_diff_moves(aapl, timespan=timespan, column_slice=["Close", "Volume"], fillna=True, merge_result=True)

Date        Open        High         Low       Close     Volume  \
0     1980-12-12   28.750000   28.875000   28.750000   28.750000  117258400   
1     1980-12-15   27.375001   27.375001   27.250000   27.250000   43971200   
2     1980-12-16   25.375000   25.375000   25.250000   25.250000   26432000   
3     1980-12-17   25.875000   25.999999   25.875000   25.875000   21610400   
4     1980-12-18   26.625000   26.750000   26.625000   26.625000   18362400   
5     1980-12-19   28.249999   28.374999   28.249999   28.249999   12157600   
6     1980-12-22   29.625000   29.750000   29.625000   29.625000    9340800   
7     1980-12-23   30.874999   30.999999   30.874999   30.874999   11737600   
8     1980-12-24   32.500000   32.624999   32.500000   32.500000   12000800   
9     1980-12-26   35.500001   35.625001   35.500001   35.500001   13893600   
10    1980-12-29   36.000000   36.124999   36.000000   36.000000   23290400   
11    1980-12-30   35.249999   35.249999   35.125000   35.125000   17220000   
12    1980-12-31   34.250000   34.250000   34.125000   34.125000    8937600   
13    1981-01-02   34.499999   34.750001   34.499999   34.499999    5415200   
14    1981-01-05   33.875001   33.875001   33.750001   33.750001    8932000   
15    1981-01-06   32.375000   32.375000   32.250000   32.250000   11289600   
16    1981-01-07   30.999999   30.999999   30.874999   30.874999   13921600   
17    1981-01-08   30.375001   30.375001   30.250001   30.250001    9956800   
18    1981-01-09   31.874999   32.000001   31.874999   31.874999    5376000   
19    1981-01-12   31.874999   31.874999   31.625000   31.625000    5924800   
20    1981-01-13   30.625000   30.625000   30.500000   30.500000    5762400   
21    1981-01-14   30.625000   30.750000   30.625000   30.625000    3572800   
22    1981-01-15   31.250001   31.500000   31.250001   31.250001    3516800   
23    1981-01-16   31.125001   31.125001   30.999999   30.999999    3348800   
24    1981-01-19   32.875001   33.000001   32.875001   32.875001   10393600   
25    1981-01-20   32.000001   32.000001   31.874999   31.874999    7520800   
26    1981-01-21   32.500000   32.749999   32.500000   32.500000    3976000   
27    1981-01-22   32.875001   33.125000   32.875001   32.875001    8887200   
28    1981-01-23   32.875001   33.000001   32.749999   32.749999    2805600   
29    1981-01-26   32.375000   32.375000   32.250000   32.250000    6160000   
...          ...         ...         ...         ...         ...        ...   
9112  2017-02-01  127.029999  130.490005  127.010002  128.750000  111985000   
9113  2017-02-02  127.980003  129.389999  127.779999  128.529999   33710400   
9114  2017-02-03  128.309998  129.190002  128.160004  129.080002   24507300   
9115  2017-02-06  129.130005  130.500000  128.899994  130.289993   26845900   
9116  2017-02-07  130.539993  132.089996  130.449997  131.529999   38183800   
9117  2017-02-08  131.350006  132.220001  131.220001  132.039993   23004100   
9118  2017-02-09  131.649994  132.449997  131.119995  132.419998   28349900   
9119  2017-02-10  132.460007  132.940002  132.050003  132.119995   20065500   
9120  2017-02-13  133.080002  133.820007  132.750000  133.289993   23035400   
9121  2017-02-14  133.470001  135.089996  133.250000  135.020004   32815500   
9122  2017-02-15  135.520004  136.270004  134.619995  135.509995   35501600   
9123  2017-02-16  135.669998  135.899994  134.839996  135.350006   22118000   
9124  2017-02-17  135.100006  135.830002  135.100006  135.720001   22084500   
9125  2017-02-21  136.229996  136.750000  135.979996  136.699997   24265100   
9126  2017-02-22  136.429993  137.119995  136.110001  137.110001   20745300   
9127  2017-02-23  137.380005  137.479996  136.300003  136.529999   20704100   
9128  2017-02-24  135.910004  136.660004  135.279999  136.660004   21690900   
9129  2017-02-27  137.139999  137.440002  136.279999  136.929993   20196400   
9130  2017-02-28  137.080002  137.440002  136.699997  136.9900

In [173]:
calc_return(aapl, timespan=timespan, column_slice=["Close", "Volume"], merge_result=True)

Date        Open        High         Low       Close     Volume  \
0     1980-12-12   28.750000   28.875000   28.750000   28.750000  117258400   
1     1980-12-15   27.375001   27.375001   27.250000   27.250000   43971200   
2     1980-12-16   25.375000   25.375000   25.250000   25.250000   26432000   
3     1980-12-17   25.875000   25.999999   25.875000   25.875000   21610400   
4     1980-12-18   26.625000   26.750000   26.625000   26.625000   18362400   
5     1980-12-19   28.249999   28.374999   28.249999   28.249999   12157600   
6     1980-12-22   29.625000   29.750000   29.625000   29.625000    9340800   
7     1980-12-23   30.874999   30.999999   30.874999   30.874999   11737600   
8     1980-12-24   32.500000   32.624999   32.500000   32.500000   12000800   
9     1980-12-26   35.500001   35.625001   35.500001   35.500001   13893600   
10    1980-12-29   36.000000   36.124999   36.000000   36.000000   23290400   
11    1980-12-30   35.249999   35.249999   35.125000   35.125000   17220000   
12    1980-12-31   34.250000   34.250000   34.125000   34.125000    8937600   
13    1981-01-02   34.499999   34.750001   34.499999   34.499999    5415200   
14    1981-01-05   33.875001   33.875001   33.750001   33.750001    8932000   
15    1981-01-06   32.375000   32.375000   32.250000   32.250000   11289600   
16    1981-01-07   30.999999   30.999999   30.874999   30.874999   13921600   
17    1981-01-08   30.375001   30.375001   30.250001   30.250001    9956800   
18    1981-01-09   31.874999   32.000001   31.874999   31.874999    5376000   
19    1981-01-12   31.874999   31.874999   31.625000   31.625000    5924800   
20    1981-01-13   30.625000   30.625000   30.500000   30.500000    5762400   
21    1981-01-14   30.625000   30.750000   30.625000   30.625000    3572800   
22    1981-01-15   31.250001   31.500000   31.250001   31.250001    3516800   
23    1981-01-16   31.125001   31.125001   30.999999   30.999999    3348800   
24    1981-01-19   32.875001   33.000001   32.875001   32.875001   10393600   
25    1981-01-20   32.000001   32.000001   31.874999   31.874999    7520800   
26    1981-01-21   32.500000   32.749999   32.500000   32.500000    3976000   
27    1981-01-22   32.875001   33.125000   32.875001   32.875001    8887200   
28    1981-01-23   32.875001   33.000001   32.749999   32.749999    2805600   
29    1981-01-26   32.375000   32.375000   32.250000   32.250000    6160000   
...          ...         ...         ...         ...         ...        ...   
9111  2017-01-31  121.150002  121.389999  120.620003  121.349998   49201000   
9112  2017-02-01  127.029999  130.490005  127.010002  128.750000  111985000   
9113  2017-02-02  127.980003  129.389999  127.779999  128.529999   33710400   
9114  2017-02-03  128.309998  129.190002  128.160004  129.080002   24507300   
9115  2017-02-06  129.130005  130.500000  128.899994  130.289993   26845900   
9116  2017-02-07  130.539993  132.089996  130.449997  131.529999   38183800   
9117  2017-02-08  131.350006  132.220001  131.220001  132.039993   23004100   
9118  2017-02-09  131.649994  132.449997  131.119995  132.419998   28349900   
9119  2017-02-10  132.460007  132.940002  132.050003  132.119995   20065500   
9120  2017-02-13  133.080002  133.820007  132.750000  133.289993   23035400   
9121  2017-02-14  133.470001  135.089996  133.250000  135.020004   32815500   
9122  2017-02-15  135.520004  136.270004  134.619995  135.509995   35501600   
9123  2017-02-16  135.669998  135.899994  134.839996  135.350006   22118000   
9124  2017-02-17  135.100006  135.830002  135.100006  135.720001   22084500   
9125  2017-02-21  136.229996  136.750000  135.979996  136.699997   24265100   
9126  2017-02-22  136.429993  137.119995  136.110001  137.110001   20745300   
9127  2017-02-23  137.380005  137.479996  136.300003  136.529999   20704100   
9128  2017-02-24  135.910004  136.660004  135.279999  136.660004   21690900   
9129  2017-02-27  137.139999  137.440002  136.279999  136.9299

In [174]:
calc_sma(aapl, timespan, ["Close", "Volume"], merge_result=True)

Date        Open        High         Low       Close     Volume  \
0     1980-12-12   28.750000   28.875000   28.750000   28.750000  117258400   
1     1980-12-15   27.375001   27.375001   27.250000   27.250000   43971200   
2     1980-12-16   25.375000   25.375000   25.250000   25.250000   26432000   
3     1980-12-17   25.875000   25.999999   25.875000   25.875000   21610400   
4     1980-12-18   26.625000   26.750000   26.625000   26.625000   18362400   
5     1980-12-19   28.249999   28.374999   28.249999   28.249999   12157600   
6     1980-12-22   29.625000   29.750000   29.625000   29.625000    9340800   
7     1980-12-23   30.874999   30.999999   30.874999   30.874999   11737600   
8     1980-12-24   32.500000   32.624999   32.500000   32.500000   12000800   
9     1980-12-26   35.500001   35.625001   35.500001   35.500001   13893600   
10    1980-12-29   36.000000   36.124999   36.000000   36.000000   23290400   
11    1980-12-30   35.249999   35.249999   35.125000   35.125000   17220000   
12    1980-12-31   34.250000   34.250000   34.125000   34.125000    8937600   
13    1981-01-02   34.499999   34.750001   34.499999   34.499999    5415200   
14    1981-01-05   33.875001   33.875001   33.750001   33.750001    8932000   
15    1981-01-06   32.375000   32.375000   32.250000   32.250000   11289600   
16    1981-01-07   30.999999   30.999999   30.874999   30.874999   13921600   
17    1981-01-08   30.375001   30.375001   30.250001   30.250001    9956800   
18    1981-01-09   31.874999   32.000001   31.874999   31.874999    5376000   
19    1981-01-12   31.874999   31.874999   31.625000   31.625000    5924800   
20    1981-01-13   30.625000   30.625000   30.500000   30.500000    5762400   
21    1981-01-14   30.625000   30.750000   30.625000   30.625000    3572800   
22    1981-01-15   31.250001   31.500000   31.250001   31.250001    3516800   
23    1981-01-16   31.125001   31.125001   30.999999   30.999999    3348800   
24    1981-01-19   32.875001   33.000001   32.875001   32.875001   10393600   
25    1981-01-20   32.000001   32.000001   31.874999   31.874999    7520800   
26    1981-01-21   32.500000   32.749999   32.500000   32.500000    3976000   
27    1981-01-22   32.875001   33.125000   32.875001   32.875001    8887200   
28    1981-01-23   32.875001   33.000001   32.749999   32.749999    2805600   
29    1981-01-26   32.375000   32.375000   32.250000   32.250000    6160000   
...          ...         ...         ...         ...         ...        ...   
9111  2017-01-31  121.150002  121.389999  120.620003  121.349998   49201000   
9112  2017-02-01  127.029999  130.490005  127.010002  128.750000  111985000   
9113  2017-02-02  127.980003  129.389999  127.779999  128.529999   33710400   
9114  2017-02-03  128.309998  129.190002  128.160004  129.080002   24507300   
9115  2017-02-06  129.130005  130.500000  128.899994  130.289993   26845900   
9116  2017-02-07  130.539993  132.089996  130.449997  131.529999   38183800   
9117  2017-02-08  131.350006  132.220001  131.220001  132.039993   23004100   
9118  2017-02-09  131.649994  132.449997  131.119995  132.419998   28349900   
9119  2017-02-10  132.460007  132.940002  132.050003  132.119995   20065500   
9120  2017-02-13  133.080002  133.820007  132.750000  133.289993   23035400   
9121  2017-02-14  133.470001  135.089996  133.250000  135.020004   32815500   
9122  2017-02-15  135.520004  136.270004  134.619995  135.509995   35501600   
9123  2017-02-16  135.669998  135.899994  134.839996  135.350006   22118000   
9124  2017-02-17  135.100006  135.830002  135.100006  135.720001   22084500   
9125  2017-02-21  136.229996  136.750000  135.979996  136.699997   24265100   
9126  2017-02-22  136.429993  137.119995  136.110001  137.110001   20745300   
9127  2017-02-23  137.380005  137.479996  136.300003  136.529999   20704100   
9128  2017-02-24  135.910004  136.660004  135.279999  136.660004   21690900   
9129  2017-02-27  137.139999  137.440002  136.279999  136.9299

In [175]:
calc_bollinger(aapl, timespan, ["Close", "Volume"], merge_result=True, scaler=2)

Date        Open        High         Low       Close     Volume  \
0     1980-12-12   28.750000   28.875000   28.750000   28.750000  117258400   
1     1980-12-15   27.375001   27.375001   27.250000   27.250000   43971200   
2     1980-12-16   25.375000   25.375000   25.250000   25.250000   26432000   
3     1980-12-17   25.875000   25.999999   25.875000   25.875000   21610400   
4     1980-12-18   26.625000   26.750000   26.625000   26.625000   18362400   
5     1980-12-19   28.249999   28.374999   28.249999   28.249999   12157600   
6     1980-12-22   29.625000   29.750000   29.625000   29.625000    9340800   
7     1980-12-23   30.874999   30.999999   30.874999   30.874999   11737600   
8     1980-12-24   32.500000   32.624999   32.500000   32.500000   12000800   
9     1980-12-26   35.500001   35.625001   35.500001   35.500001   13893600   
10    1980-12-29   36.000000   36.124999   36.000000   36.000000   23290400   
11    1980-12-30   35.249999   35.249999   35.125000   35.125000   17220000   
12    1980-12-31   34.250000   34.250000   34.125000   34.125000    8937600   
13    1981-01-02   34.499999   34.750001   34.499999   34.499999    5415200   
14    1981-01-05   33.875001   33.875001   33.750001   33.750001    8932000   
15    1981-01-06   32.375000   32.375000   32.250000   32.250000   11289600   
16    1981-01-07   30.999999   30.999999   30.874999   30.874999   13921600   
17    1981-01-08   30.375001   30.375001   30.250001   30.250001    9956800   
18    1981-01-09   31.874999   32.000001   31.874999   31.874999    5376000   
19    1981-01-12   31.874999   31.874999   31.625000   31.625000    5924800   
20    1981-01-13   30.625000   30.625000   30.500000   30.500000    5762400   
21    1981-01-14   30.625000   30.750000   30.625000   30.625000    3572800   
22    1981-01-15   31.250001   31.500000   31.250001   31.250001    3516800   
23    1981-01-16   31.125001   31.125001   30.999999   30.999999    3348800   
24    1981-01-19   32.875001   33.000001   32.875001   32.875001   10393600   
25    1981-01-20   32.000001   32.000001   31.874999   31.874999    7520800   
26    1981-01-21   32.500000   32.749999   32.500000   32.500000    3976000   
27    1981-01-22   32.875001   33.125000   32.875001   32.875001    8887200   
28    1981-01-23   32.875001   33.000001   32.749999   32.749999    2805600   
29    1981-01-26   32.375000   32.375000   32.250000   32.250000    6160000   
...          ...         ...         ...         ...         ...        ...   
9111  2017-01-31  121.150002  121.389999  120.620003  121.349998   49201000   
9112  2017-02-01  127.029999  130.490005  127.010002  128.750000  111985000   
9113  2017-02-02  127.980003  129.389999  127.779999  128.529999   33710400   
9114  2017-02-03  128.309998  129.190002  128.160004  129.080002   24507300   
9115  2017-02-06  129.130005  130.500000  128.899994  130.289993   26845900   
9116  2017-02-07  130.539993  132.089996  130.449997  131.529999   38183800   
9117  2017-02-08  131.350006  132.220001  131.220001  132.039993   23004100   
9118  2017-02-09  131.649994  132.449997  131.119995  132.419998   28349900   
9119  2017-02-10  132.460007  132.940002  132.050003  132.119995   20065500   
9120  2017-02-13  133.080002  133.820007  132.750000  133.289993   23035400   
9121  2017-02-14  133.470001  135.089996  133.250000  135.020004   32815500   
9122  2017-02-15  135.520004  136.270004  134.619995  135.509995   35501600   
9123  2017-02-16  135.669998  135.899994  134.839996  135.350006   22118000   
9124  2017-02-17  135.100006  135.830002  135.100006  135.720001   22084500   
9125  2017-02-21  136.229996  136.750000  135.979996  136.699997   24265100   
9126  2017-02-22  136.429993  137.119995  136.110001  137.110001   20745300   
9127  2017-02-23  137.380005  137.479996  136.300003  136.529999   20704100   
9128  2017-02-24  135.910004  136.660004  135.279999  136.660004   21690900   
9129  2017-02-27  137.139999  137.440002  136.279999  136.9299

In [176]:
sp500 = load_raw_frame("^GSPC")

In [177]:
sp500_i = sp500.set_index("Date")
aapl_i = aapl.set_index("Date")

In [178]:
joined_df = pd.merge(aapl_i.loc[:,["Close"]], sp500_i.loc[:,["Close"]], how="left", left_index=True, right_index=True)

In [180]:
beta_df = timewindow_alphabeta(aapl, sp500, ["Close", "Open"], 10, merge_result=False)
beta_df.head()

timewindow_beta_10_Close  timewindow_alpha_10_Close  \
Date                                                              
1980-12-12                       0.0                        0.0   
1980-12-15                       0.0                        0.0   
1980-12-16                       0.0                        0.0   
1980-12-17                       0.0                        0.0   
1980-12-18                       0.0                        0.0   

            timewindow_beta_10_Open  timewindow_alpha_10_Open  
Date                                                           
1980-12-12                      0.0                       0.0  
1980-12-15                      0.0                       0.0  
1980-12-16                      0.0                       0.0  
1980-12-17                      0.0                       0.0  
1980-12-18                      0.0                       0.0

In [181]:
beta_df

timewindow_beta_10_Close  timewindow_alpha_10_Close  \
Date                                                              
1980-12-12                  0.000000                   0.000000   
1980-12-15                  0.000000                   0.000000   
1980-12-16                  0.000000                   0.000000   
1980-12-17                  0.000000                   0.000000   
1980-12-18                  0.000000                   0.000000   
1980-12-19                  0.000000                   0.000000   
1980-12-22                  0.000000                   0.000000   
1980-12-23                  0.000000                   0.000000   
1980-12-24                  0.000000                   0.000000   
1980-12-26                  0.602909                 115.725489   
1980-12-29                  0.462633                 120.045095   
1980-12-30                  0.362929                 123.315977   
1980-12-31                  0.272490                 126.354180   
1981-01-02                  0.251707                 127.135722   
1981-01-05                  0.193142                 129.387471   
1981-01-06                 -0.035484                 137.394066   
1981-01-07                 -0.001231                 136.179306   
1981-01-08                  0.225218                 128.372000   
1981-01-09                  0.319368                 124.999134   
1981-01-12                  0.368983                 123.178736   
1981-01-13                  0.629617                 114.740313   
1981-01-14                  0.873622                 107.020322   
1981-01-15                  1.018467                 102.518674   
1981-01-16                  1.410476                  90.409048   
1981-01-19                  0.828619                 108.391856   
1981-01-20                 -0.015581                 134.178299   
1981-01-21                 -0.262831                 141.581753   
1981-01-22                 -0.851694                 160.037691   
1981-01-23                 -1.096102                 167.556352   
1981-01-26                 -1.198070                 170.504536   
...                              ...                        ...   
2017-01-31                 11.642928                 873.037655   
2017-02-01                  1.251145                2128.627324   
2017-02-02                  0.294377                2246.556085   
2017-02-03                  0.572523                2214.544241   
2017-02-06                  0.158967                2268.198686   
2017-02-07                  0.047792                2283.289855   
2017-02-08                  0.451859                2231.672412   
2017-02-09                  1.211181                2135.309142   
2017-02-10                  2.196812                2009.294779   
2017-02-13                  3.356260                1860.809625   
2017-02-14                  8.176562                1229.153197   
2017-02-15                  8.923136                1132.060825   
2017-02-16                  9.587602                1044.460842   
2017-02-17                 11.906598                 734.279305   
2017-02-21                 13.008930                 586.249738   
2017-02-22                 12.059992                 713.623372   
2017-02-23                 11.086410                 846.533664   
2017-02-24                 10.527539                 923.451696   
2017-02-27                 11.473788                 795.172752   
2017-02-28                 13.083336                 575.162225   
2017-03-01                 10.819780                 884.253245   
2017-03-02                 10.032198                 991.744418   
2017-03-03                  8.439390                1209.928334   
2017-03-06                  7.047734                1401.168434   
2017-03-07                  5.901087                1557.903858   
2017-03-08                  5.031013                1677.179407   
2017-03-09                  4.908720                1693.14

In [182]:
beta_df

timewindow_beta_10_Close  timewindow_alpha_10_Close  \
Date                                                              
1980-12-12                  0.000000                   0.000000   
1980-12-15                  0.000000                   0.000000   
1980-12-16                  0.000000                   0.000000   
1980-12-17                  0.000000                   0.000000   
1980-12-18                  0.000000                   0.000000   
1980-12-19                  0.000000                   0.000000   
1980-12-22                  0.000000                   0.000000   
1980-12-23                  0.000000                   0.000000   
1980-12-24                  0.000000                   0.000000   
1980-12-26                  0.602909                 115.725489   
1980-12-29                  0.462633                 120.045095   
1980-12-30                  0.362929                 123.315977   
1980-12-31                  0.272490                 126.354180   
1981-01-02                  0.251707                 127.135722   
1981-01-05                  0.193142                 129.387471   
1981-01-06                 -0.035484                 137.394066   
1981-01-07                 -0.001231                 136.179306   
1981-01-08                  0.225218                 128.372000   
1981-01-09                  0.319368                 124.999134   
1981-01-12                  0.368983                 123.178736   
1981-01-13                  0.629617                 114.740313   
1981-01-14                  0.873622                 107.020322   
1981-01-15                  1.018467                 102.518674   
1981-01-16                  1.410476                  90.409048   
1981-01-19                  0.828619                 108.391856   
1981-01-20                 -0.015581                 134.178299   
1981-01-21                 -0.262831                 141.581753   
1981-01-22                 -0.851694                 160.037691   
1981-01-23                 -1.096102                 167.556352   
1981-01-26                 -1.198070                 170.504536   
...                              ...                        ...   
2017-01-31                 11.642928                 873.037655   
2017-02-01                  1.251145                2128.627324   
2017-02-02                  0.294377                2246.556085   
2017-02-03                  0.572523                2214.544241   
2017-02-06                  0.158967                2268.198686   
2017-02-07                  0.047792                2283.289855   
2017-02-08                  0.451859                2231.672412   
2017-02-09                  1.211181                2135.309142   
2017-02-10                  2.196812                2009.294779   
2017-02-13                  3.356260                1860.809625   
2017-02-14                  8.176562                1229.153197   
2017-02-15                  8.923136                1132.060825   
2017-02-16                  9.587602                1044.460842   
2017-02-17                 11.906598                 734.279305   
2017-02-21                 13.008930                 586.249738   
2017-02-22                 12.059992                 713.623372   
2017-02-23                 11.086410                 846.533664   
2017-02-24                 10.527539                 923.451696   
2017-02-27                 11.473788                 795.172752   
2017-02-28                 13.083336                 575.162225   
2017-03-01                 10.819780                 884.253245   
2017-03-02                 10.032198                 991.744418   
2017-03-03                  8.439390                1209.928334   
2017-03-06                  7.047734                1401.168434   
2017-03-07                  5.901087                1557.903858   
2017-03-08                  5.031013                1677.179407   
2017-03-09                  4.908720                1693.14

In [183]:
pd.merge(aapl,beta_df,left_on='Date',right_index=True)

Date        Open        High         Low       Close     Volume  \
0     1980-12-12   28.750000   28.875000   28.750000   28.750000  117258400   
1     1980-12-15   27.375001   27.375001   27.250000   27.250000   43971200   
2     1980-12-16   25.375000   25.375000   25.250000   25.250000   26432000   
3     1980-12-17   25.875000   25.999999   25.875000   25.875000   21610400   
4     1980-12-18   26.625000   26.750000   26.625000   26.625000   18362400   
5     1980-12-19   28.249999   28.374999   28.249999   28.249999   12157600   
6     1980-12-22   29.625000   29.750000   29.625000   29.625000    9340800   
7     1980-12-23   30.874999   30.999999   30.874999   30.874999   11737600   
8     1980-12-24   32.500000   32.624999   32.500000   32.500000   12000800   
9     1980-12-26   35.500001   35.625001   35.500001   35.500001   13893600   
10    1980-12-29   36.000000   36.124999   36.000000   36.000000   23290400   
11    1980-12-30   35.249999   35.249999   35.125000   35.125000   17220000   
12    1980-12-31   34.250000   34.250000   34.125000   34.125000    8937600   
13    1981-01-02   34.499999   34.750001   34.499999   34.499999    5415200   
14    1981-01-05   33.875001   33.875001   33.750001   33.750001    8932000   
15    1981-01-06   32.375000   32.375000   32.250000   32.250000   11289600   
16    1981-01-07   30.999999   30.999999   30.874999   30.874999   13921600   
17    1981-01-08   30.375001   30.375001   30.250001   30.250001    9956800   
18    1981-01-09   31.874999   32.000001   31.874999   31.874999    5376000   
19    1981-01-12   31.874999   31.874999   31.625000   31.625000    5924800   
20    1981-01-13   30.625000   30.625000   30.500000   30.500000    5762400   
21    1981-01-14   30.625000   30.750000   30.625000   30.625000    3572800   
22    1981-01-15   31.250001   31.500000   31.250001   31.250001    3516800   
23    1981-01-16   31.125001   31.125001   30.999999   30.999999    3348800   
24    1981-01-19   32.875001   33.000001   32.875001   32.875001   10393600   
25    1981-01-20   32.000001   32.000001   31.874999   31.874999    7520800   
26    1981-01-21   32.500000   32.749999   32.500000   32.500000    3976000   
27    1981-01-22   32.875001   33.125000   32.875001   32.875001    8887200   
28    1981-01-23   32.875001   33.000001   32.749999   32.749999    2805600   
29    1981-01-26   32.375000   32.375000   32.250000   32.250000    6160000   
...          ...         ...         ...         ...         ...        ...   
9111  2017-01-31  121.150002  121.389999  120.620003  121.349998   49201000   
9112  2017-02-01  127.029999  130.490005  127.010002  128.750000  111985000   
9113  2017-02-02  127.980003  129.389999  127.779999  128.529999   33710400   
9114  2017-02-03  128.309998  129.190002  128.160004  129.080002   24507300   
9115  2017-02-06  129.130005  130.500000  128.899994  130.289993   26845900   
9116  2017-02-07  130.539993  132.089996  130.449997  131.529999   38183800   
9117  2017-02-08  131.350006  132.220001  131.220001  132.039993   23004100   
9118  2017-02-09  131.649994  132.449997  131.119995  132.419998   28349900   
9119  2017-02-10  132.460007  132.940002  132.050003  132.119995   20065500   
9120  2017-02-13  133.080002  133.820007  132.750000  133.289993   23035400   
9121  2017-02-14  133.470001  135.089996  133.250000  135.020004   32815500   
9122  2017-02-15  135.520004  136.270004  134.619995  135.509995   35501600   
9123  2017-02-16  135.669998  135.899994  134.839996  135.350006   22118000   
9124  2017-02-17  135.100006  135.830002  135.100006  135.720001   22084500   
9125  2017-02-21  136.229996  136.750000  135.979996  136.699997   24265100   
9126  2017-02-22  136.429993  137.119995  136.110001  137.110001   20745300   
9127  2017-02-23  137.380005  137.479996  136.300003  136.529999   20704100   
9128  2017-02-24  135.910004  136.660004  135.279999  136.660004   21690900   
9129  2017-02-27  137.139999  137.440002  136.279999  136.9299

In [ ]:
"""
1980-12-26     0.602909
1980-12-29     0.462633
1980-12-30     0.362929
1980-12-31     0.272490
1981-01-02     0.251707
1981-01-05     0.193142
"""

In [ ]:
_i = 10

s = joined_df.iloc[ 0:10, :]["Close_x"]
m = joined_df.iloc[ 0:10, :]["Close_y"]

slope, intercept, r_value, p_value, std_err = stats.linregress(s.values,m.values)
slope

In [ ]:
tier1_df = calc_measures_tier1(aapl)
tier1_df

In [202]:
tier1_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9141 entries, 0 to 9140
Columns: 159 entries, Date to timewindow_alpha_20_Low
dtypes: float64(157), int64(1), object(1)
memory usage: 11.2+ MB


In [ ]:
store_tier1_frame(tier1_df, "AAPL")

In [ ]:
load_tier1_frame("AAPL")

In [ ]:
tier1_df = calc_measures_tier1(aapl, verbose=True)

In [ ]:
%%timeit
tier1_df = calc_measures_tier1(aapl, verbose=True)

In [ ]:
aapl.head()

In [ ]:
timewindow_alpha(aapl, sp500, ["Close","Volume","Low"], 10)

In [ ]:
timewindow_beta(aapl, sp500, ["Close","Volume","Low"], 10)

In [ ]:
timewindow_alphabeta(aapl, sp500, ["Close","Volume","Low"], 10)

In [ ]:
ticker_list = pd.read_csv("RES_initial_dataload.csv")
ticker_list = ticker_list[ticker_list["STATUS"] == "OK"].iloc[0]
ticker_list

for ix, row in ticker_list.iterrows():
    itr_ticker = row["Symbol"]
    print("\n\n - {} - \n".format(itr_ticker)
    itr_df = load_raw_frame("AAPL")
    tier1_df = calc_measures_tier1(itr_df, verbose=True)
    store_tier1_frame(tier1_df, itr_ticker)
    

In [ ]:
aapl = load_raw_frame("AAPL")
tier1_df = calc_measures_tier1(aapl, verbose=True)
store_tier1_frame(tier1_df, "APPL")

In [ ]:
tier1_df.head()

In [ ]:
# Open Close Diff             |   Open(t) - Close(t -1)
# Open Close Return           |   (Close(t -1) / Open(t)) - 1
# Open Close Return MA        |   SMA(Open Close Return)
# Open Close Return Bollinger |   Bollinger(Open Close Return)

In [ ]:
aapl.head()

In [ ]:
# Open Close Diff

aftermarketreturn = (aapl["Open"] - aapl["Close"].shift(1))
aftermarketreturn.head()

In [ ]:
# Open Close Return

aftermarketdiff = (aapl["Open"] / aapl["Close"].shift(1)) - 1
aftermarketdiff.head()

In [199]:
calc_aftermarket_bollinger??

In [ ]:
# Open Close Diff             |   Open(t) - Close(t -1)
calc_aftermarket_diff(aapl).head()

In [ ]:
# Open Close Return           |   (Close(t -1) / Open(t)) - 1
calc_aftermarket_return(aapl).head()

In [ ]:
not "D" in aapl.columns

In [ ]:
# Open Close Return MA        |   SMA(Open Close Return)
calc_aftermarket_sma(aapl, timespan)

In [ ]:
# Open Close Return Bollinger |   Bollinger(Open Close Return)
calc_aftermarket_bollinger(aapl, timespan)

In [ ]:
aapl_t1 = load_tier1_frame("AAPL")

In [ ]:
aapl_t1.columns.tolist()

In [ ]:
returns_cols = [x for x in aapl_t1.columns.tolist() if "timewindow_return_" in x]
returns_cols

In [ ]:
returns_cols_disc = ["discrete_{}".format(x) for x in returns_cols]

In [ ]:
data = np.random.random(100)
digitized = np.digitize(data, bins)
bin_means = [data[digitized == i].mean() for i in range(1, len(bins))]


In [ ]:
x = np.array(np.linspace(0, 0.9, 10))
bins = np.array([-2.0, -1.0, -0.8, -0.5, -0.3, -0.1, -0.05, 0.01, 0.05, 0.07, 0.1, 0.2, 0.4, 1.0])
inds = np.digitize(x, bins)


for n in range(x.size):
    print(bins[inds[n]-1], "<=", x[n], "<", bins[inds[n]])
    


In [ ]:
x = np.array(np.linspace(0, 0.9, 10))
bins = np.array([0.0, 0.01, 0.05, 0.07, 0.1, 0.2, 0.4, 1.0])
inds = np.digitize(x, bins)


for n in range(x.size):
    print(bins[inds[n]-1], "<=", x[n], "<", bins[inds[n]])
    


In [ ]:
x = np.array(np.linspace(-1.9, 0.0, 10))
bins = np.array([-2.0, -1.0, -0.8, -0.5, -0.3, -0.1, -0.05, 0.0, 0.001])
inds = np.digitize(x, bins)


for n in range(x.size):
    print(bins[inds[n]-1], "<=", x[n], "<", bins[inds[n]])
    


In [ ]:
x = np.array(np.linspace(-2.1, 1.0, 10))
bins = np.array([-2.0, -1.0, -0.8, -0.5, -0.3, -0.1, -0.05, 0.01, 0.05, 0.07, 0.1, 0.2, 0.4, 1.0])
inds = np.digitize(x, bins)
inds

In [ ]:
np.clip(inds, 0, bins.shape[0] - 1)

In [ ]:
x = aapl_t1.loc[ : , returns_cols].fillna(0).values
bins = np.array([-2.0, -1.0, -0.8, -0.5, -0.3, -0.1, -0.05, 0.01, 0.05, 0.07, 0.1, 0.2, 0.4, 1.0])
inds = np.digitize(x, bins, right=False)
inds = np.clip(inds, 0, bins.shape[0] - 1)

discretized_df = pd.DataFrame(inds, columns=returns_cols_disc)
discretized_df.head()

In [ ]:
pd.DataFrame(bins, columns=["Values"])

In [ ]:
loss_bins = np.array([-2.0, -1.0, -0.5, 0.0])
gain_bins = np.array([0.0, 0.05, 1.0, 2.0])

In [ ]:
loss_bins

In [ ]:
gain_bins

In [ ]:
x = np.linspace(-2.5, 2.5, 7)
x

In [ ]:
inds_loss = np.digitize(x, loss_bins, right=False)
inds_loss = np.clip(inds_loss, 0, loss_bins.shape[0] - 1)
inds_loss

In [ ]:
inds_gain = np.digitize(x, gain_bins)
inds_gain = np.clip(inds_gain, 0, gain_bins.shape[0] - 1)
inds_gain

In [ ]:
bins = np.array([-2.0, -1.0, -0.8, -0.5, -0.3, -0.1, -0.05, 0.0, 0.01, 0.05, 0.07, 0.1, 0.2, 0.4, 1.0])

In [ ]:
def discretize(bins, return_index, v):
    _r = None
    if v <= 0:
        inds_gain = np.digitize(v, bins, right=True)
        inds_gain = np.clip(inds_gain, 0, bins.shape[0] - 1)

        _r = inds_gain
    else:
        inds_gain = np.digitize(v, bins, right=False) - 1
        inds_gain = np.clip(inds_gain, 0, bins.shape[0] - 1)

        _r = inds_gain
    
    if not ( return_index ):
        _r = bins[_r]
    
    return _r

_bins = np.array([-2.0, -1.0, -0.8, -0.5, -0.3, -0.1, -0.05, 0.0, 0.01, 0.05, 0.07, 0.1, 0.2, 0.4, 1.0])
_disc = partial(discretize, _bins, False)

_real = [-3, -2, -1, -0.7, -0.5, 0, 0.005, 0.01, 0.012, 0.04, 0.05, 0.06, 0.08, 0.1, 0.15, 0.2, 0.3, 0.4, 0.9, 1, 1.2, 40]
_discrete = pd.DataFrame(_real, index=_real).applymap(_disc)



In [ ]:
x = aapl_t1.loc[ : , returns_cols].fillna(0)

discretize_df = x.applymap(discretize)
discretized_df = pd.DataFrame(discretize_df, columns=returns_cols_disc)
discretized_df

In [208]:
tier1_df = calc_measures_tier1(aapl)
print(tier1_df.columns.shape)
tier1_df = calc_measures_tier2(tier1_df)
print(tier1_df.columns.shape)

| | START     - 2017-03-15 16:15:44.192986
\ / SMA       - 0:00:00.086125
\ / RETURNS   - 0:00:00.092693
\ / DIFF MOVE - 0:00:00.096596
\ / BOLLINGER - 0:00:00.117711


/opt/conda/lib/python3.5/site-packages/scipy/stats/_stats_mstats_common.py:95: RuntimeWarning: invalid value encountered in double_scalars
  slope = r_num / ssxm
/opt/conda/lib/python3.5/site-packages/scipy/stats/_stats_mstats_common.py:97: RuntimeWarning: divide by zero encountered in double_scalars
  sterrest = np.sqrt((1 - r**2) * ssym / ssxm / df)


\ / ALPHABETA - 0:02:26.536868
 V  END       - 2017-03-15 16:18:11.122979 (TOOK 0:02:26.929993)
(161,)
| | START     - 2017-03-15 16:18:11.123713
\ / AM DIFF       - 0:00:00.001957
\ / AM RETURNS   - 0:00:00.001845
\ / AM SMA - 0:00:00.047715
\ / AM BOLLINGER - 0:00:00.065953
 V  END       - 2017-03-15 16:18:11.241268 (TOOK 0:00:00.117555)
(184,)


In [31]:
aapl = load_raw_frame("AAPL")
aapl["Date"].head()

0   1980-12-12
1   1980-12-15
2   1980-12-16
3   1980-12-17
4   1980-12-18
Name: Date, dtype: datetime64[ns]

In [44]:
from dateutil import parser as dtparser

In [50]:
(aapl["Date"] > dtparser.parse("12-12-1980")) & (aapl["Date"] <= dtparser.parse("16-12-1980"))

0       False
1        True
2        True
3       False
4       False
5       False
6       False
7       False
8       False
9       False
10      False
11      False
12      False
13      False
14      False
15      False
16      False
17      False
18      False
19      False
20      False
21      False
22      False
23      False
24      False
25      False
26      False
27      False
28      False
29      False
        ...  
9112    False
9113    False
9114    False
9115    False
9116    False
9117    False
9118    False
9119    False
9120    False
9121    False
9122    False
9123    False
9124    False
9125    False
9126    False
9127    False
9128    False
9129    False
9130    False
9131    False
9132    False
9133    False
9134    False
9135    False
9136    False
9137    False
9138    False
9139    False
9140    False
9141    False
Name: Date, dtype: bool